firestore/integration_test.go

```diff
@@ -506,6 +506,142 @@ func TestIntegration_GetAll(t *testing.T) {
 	})
 }
 
+type runWithOptionsTestcase struct {
+	desc               string
+	wantExplainMetrics *ExplainMetrics
+	wantSnapshots      bool
+	opts               []RunOption
+}
+
+func getRunWithOptionsTestcases(t *testing.T) ([]runWithOptionsTestcase, []*DocumentRef) {
+	type getAll struct{ N int }
+
+	count := 5
+	h := testHelper{t}
+	coll := integrationColl(t)
+	var wantDocRefs []*DocumentRef
+	for i := 0; i < count; i++ {
+		doc := coll.Doc("getRunWithOptionsTestcases" + fmt.Sprint(i))
+		wantDocRefs = append(wantDocRefs, doc)
+		h.mustCreate(doc, getAll{N: i})
+	}
+
+	wantPlanSummary := &PlanSummary{
+		IndexesUsed: []*map[string]interface{}{
+			{
+				"properties":  "(__name__ ASC)",
+				"query_scope": "Collection",
+			},
+		},
+	}
+	return []runWithOptionsTestcase{
+		{
+			desc:          "No ExplainOptions",
+			wantSnapshots: true,
+		},
+
+		{
+			desc: "ExplainOptions.Analyze is false",
+			opts: []RunOption{ExplainOptions{}},
+			wantExplainMetrics: &ExplainMetrics{
+				PlanSummary: wantPlanSummary,
+			},
+		},
+		{
+			desc: "ExplainOptions.Analyze is true",
+			opts: []RunOption{ExplainOptions{Analyze: true}},
+			wantExplainMetrics: &ExplainMetrics{
+				ExecutionStats: &ExecutionStats{
+					ReadOperations:  int64(count),
+					ResultsReturned: int64(count),
+					DebugStats: &map[string]interface{}{
+						"documents_scanned":     fmt.Sprint(count),
+						"index_entries_scanned": fmt.Sprint(count),
+					},
+				},
+				PlanSummary: wantPlanSummary,
+			},
+			wantSnapshots: true,
+		},
+	}, wantDocRefs
+}
+
+func TestIntegration_GetAllWithOptions(t *testing.T) {
+	coll := integrationColl(t)
+	ctx := context.Background()
+	testcases, wantDocRefs := getRunWithOptionsTestcases(t)
+	snapshotRefIDs := []string{}
+	for _, wantRef := range wantDocRefs {
+		snapshotRefIDs = append(snapshotRefIDs, wantRef.ID)
+	}
+
+	defer func() {
+		deleteDocuments(wantDocRefs)
+	}()
+
+	for _, testcase := range testcases {
+		gotResult, gotErr := coll.Documents(ctx).GetAllWithOptions(testcase.opts...)
+		if gotErr != nil {
+			t.Fatalf("%v: err: got: %+v, want: nil", testcase.desc, gotErr)
+		}
+
+		if gotResult != nil {
+			gotIDs := []string{}
+			for _, gotSnapshot := range gotResult.DocumentSnapshots {
+				gotIDs = append(gotIDs, gotSnapshot.Ref.ID)
+			}
+
+			if (testcase.wantSnapshots && !testutil.Equal(gotIDs, snapshotRefIDs)) || (!testcase.wantSnapshots && len(gotIDs) != 0) {
+				t.Errorf("%v: snapshots ID: got: %+v, want: %+v", testcase.desc, gotIDs, snapshotRefIDs)
+			}
+
+			if err := cmpExplainMetrics(gotResult.ExplainMetrics, testcase.wantExplainMetrics); err != nil {
+				t.Errorf("%v: %+v", testcase.desc, err)
+			}
+		}
+
+	}
+}
+
+func TestIntegration_QueryRunOptions(t *testing.T) {
```

Comment:
can we add some tests in transactions as well?

Reply:
Added